text/0000-guaranteed-tco.md

```diff
@@ -0,0 +1,602 @@
+- Feature Name: guaranteed_tco
```

Comment:
I agree "explicit" would be better than "guaranteed" as it does not imply to be something general. Though I'm not so sure about the tail calls part. A tail call is only that the final action of a function is to call another function, it does not imply any optimization which *is* the primary purpose of the RFC. For this reason I'm also a bit hesitant to update the text as you suggest. Regarding searchability I suspect there is not much difference (I'm pretty sure I searched for both when I wanted this feature).

What do you think of `explicit_tco` or `explicit_tail_call_optimization`?

Reply:
> A tail call is only that the final action of a function is to call another function, it does not imply any optimization which _is_ the primary purpose of the RFC.

This RFC cannot literally *guarantee* tail calls being optimized in the first place, as TCO is an optimization in the first place. The important thing that the RFC does is to pave the way for TCO to actually be applied as an optimization, rather than getting trivially blocked in the majority of cases by pesky drop order considerations.

Reply:
The intent of `become` is that it guarantees O(1) stack usage instead of O(n) for n tail calls. This is guaranteed in both debug and release builds, and it is a compiler bug if this does not happen.

Reply:
There is also another option for the feature name, see this [comment](https://github.com/rust-lang/rfcs/pull/3407#issuecomment-1501027944):

> I recommend that you call this "Proper Tail Calls" and not "Guaranteed TCO".
> 
>     * The term "proper tail call" (PTC) is a standard tech term for the feature.
> 
>     * The purpose of PTC is to generate a jump to another function, instead of pushing a new stack frame and calling it. This ensures that a chain of tail calls consumes constant stack space. The constant stack space guarantee is the purpose of this feature. The purpose is not "optimization".
> 
>     * PTC is not always an optimization. In some cases, the code for jumping to a function can be more expensive than creating a new stack frame, because setting up new arguments on the stack while clearing out old stack arguments might be more expensive than just pushing the new arguments. There might be extra copying involved.
> 
>     * Using the word "optimization" may cause confusion. Compiler optimizations are distinct from language features that provide semantic guarantees, which is what PTC is (a guaranteed bound on space consumption). The fact that PTC is sometimes slower may cause confusion. People may point out that your "guaranteed optimization" is not always an optimization.

Reply:
The previous RFC #1888 also used "Proper Tail Calls". I chose not to use this term as it seems to imply support for general tail calls (which this RFC does not currently allow) and I think it is more a functional programming term.

That said, I don't want to hold up finding an appropriate feature name. It looks like there is consensus on not using "optimization" as part of the feature name. So I'm happy to change it to "Explicit Tail Calls" or "Proper Tail Calls".

Reply:
`tail_calls_demand` because `become` demands to compiler use tail calls

Reply:
Another name for this, alongside "proper tail calls", is "tail call elimination". That's because TCO is an optimization that the compiler may or may not perform, and TCE is a language feature (see for example [this proposal for the C language](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2920.pdf) that on page 4 contrasts TCE with TCO on the basis that TCO is a compiler optimization)

In other words, the Rust compiler already does TCO (though not on all cases), what it doesn't have is syntax for guaranteed TCE.

Because of this, maybe whenever this RFC refers to TCO it should instead refer to TCE instead.

Reply:
@VitWW also a possibility, though, I personally would prefer "Explicit Tails Calls" as it sounds more descriptive to me.

@dlight thank you for that, I have to admit I did not know there was a distinction. I also found these [slides](http://lampwww.epfl.ch/teaching/archive/advanced_compiler/2007/resources/slides/act-2007-07-tail-calls.pdf) where page 12 talks about their difference. I'll do another pass on the RFC to update where appropriate, maybe it would also be helpful to have a section dedicated to these terms.

This also makes "Tail Call Elimination" my preferred feature name. Any thoughts on this option? (@WaffleLapkin, @doug-moen)

Reply:
I did rewrite the text to use TCE instead: https://github.com/rust-lang/rfcs/pull/3407/commits/bef3933b609ea3cef87b0720aec334121ec28fde

Reply:
The use of "Tail Call Elimination" in the RFC text is definitely an improvement and a good call.

However I don't think that it is a good feature name, TCE is the process/mechanism behind tail calls and you could argue that the compiler already does TCE with TCO.

I still think that `explicit_tail_calls` (or `proper_tail_calls` as suggested by @doug-moen) are better feature names as they, in my opinion, better describe it.

Reply:
Oh, then I misunderstood the difference. I thought TCE describes a language feature and not the process as you say. In that case, I agree TCE is not a good feature name. I'll also need to do another pass on the text then.

Does anyone have some thoughts on which option would be better between: `explicit_tail_calls` and `proper_tail_calls`?

Reply:
Not sure how important naming is at this point although I also prefer `explicit_tail_calls` for the reason that this is what this RFC wants to allow in Rust.

We could maybe do an emoji poll if people think it is necessary to resolve this question now:

🎉 := `explicit_tail_calls`
🚀 := `proper_tail_calls`
💯  := `tail_call_elimination`

Reply:
Thanks @Robbepop, this is a great idea!

Also not sure how important the naming is but at some point one is needed and I already managed to pick a bad one once.

Reply:
Thanks everyone for voting!